quil-rs/src/stats/mod.rs

```diff
@@ -10,53 +10,107 @@ use std::ops::Neg;
 
 use crate::{
     instruction::{Instruction, Qubit},
+    program::graph::InstructionBlock,
     Program,
 };
 use execution_graph::{Error as ExecutionGraphError, ExecutionGraph};
 
 pub trait InstructionsSource {
+    type QubitSet;
+
+    fn is_empty(&self) -> bool;
+    fn len(&self) -> usize;
     fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_;
-    fn num_instructions(&self) -> usize;
-    fn get_used_qubits(&self) -> impl IntoIterator<Item = &Qubit>;
+    fn get_used_qubits(&self) -> &Self::QubitSet;
+}
+
+// Blanket-impl to ensure that if an owned type implements `InstructionsSource`,
+// then a reference to that type does as well
+impl<S: InstructionsSource> InstructionsSource for &S {
+    type QubitSet = S::QubitSet;
+
+    fn is_empty(&self) -> bool {
+        (*self).is_empty()
+    }
+
+    fn len(&self) -> usize {
+        (*self).len()
+    }
+
+    fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_ {
+        (*self).body_instructions()
+    }
+
+    fn get_used_qubits(&self) -> &Self::QubitSet {
+        (*self).get_used_qubits()
+    }
 }
 
 impl InstructionsSource for Program {
+    type QubitSet = HashSet<Qubit>;
+
+    fn is_empty(&self) -> bool {
+        Program::is_empty(self)
+    }
+
+    fn len(&self) -> usize {
+        Program::len(self)
+    }
+
     fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_ {
         Program::body_instructions(self)
     }
 
-    fn num_instructions(&self) -> usize {
-        Program::num_instructions(self)
+    fn get_used_qubits(&self) -> &Self::QubitSet {
+        Program::get_used_qubits(self)
     }
+}
 
-    // There's no generic trait for "set-like" types, so we return a concrete
-    // type here (as an API convenience) but do not require this type in the
-    // trait definition.
-    #[allow(refining_impl_trait)]
-    fn get_used_qubits(&self) -> &HashSet<Qubit> {
-        Program::get_used_qubits(self)
+struct InstructionBlockWithQubitSet<'a, 'b> {
+    block: &'a InstructionBlock<'b>,
+    qubits: HashSet<&'a Qubit>,
+}
+
+impl<'a, 'b> From<&'a InstructionBlock<'b>> for InstructionBlockWithQubitSet<'a, 'b> {
+    fn from(block: &'a InstructionBlock<'b>) -> Self {
+        let qubits = block
+            .instructions
+            .iter()
+            .filter_map(|i| match i {
+                Instruction::Gate(_) | Instruction::Measurement(_) => Some(i.get_qubits()),
+                _ => None,
+            })
```

Comment:
Why not use `get_qubits` on `Instruction` directly?

Reply:
Do you mean without the `filter_map`, or as a fully-qualified `Instruction::get_qubits`?